shimmingtoolbox/coils/siemens_basis.py

```diff
@@ -0,0 +1,143 @@
+                           system (i.e. DICOM reference, units of mm)
+        y (numpy.ndarray): 3-D arrays of grid coordinates (same shape as x). "Anterior->Posterior" grid coordinates in
+                           the patient coordinate system (i.e. DICOM reference, units of mm)
+
+        z (numpy.ndarray): 3-D arrays of grid coordinates (same shape as x). "Inferior->Superior" grid coordinates in
+                           the patient coordinate system (i.e. DICOM reference, units of mm)
+
+    Returns:
+        numpy.ndarray: 4-D array of spherical harmonic basis fields
+
+    NOTES:
+        For now, `orders` is, in fact, ignored: fixed as [1:2]—which is suitable for the Prisma (presumably other
+        Siemens systems as well) however, the 3rd-order shims of the Terra should ultimately be accommodated too.
+        (Requires checking the Adjustments/Shim card to see what the corresponding terms and values actually are). So,
+        for now, `basis` will always be returned with 8 terms along the 4th dim.
+    """
+
+    # Local functions
+    def reorder_to_siemens(spher_harm):
+        """
+        Reorder 1st - 2nd order basis terms along 4th dim. From
+        1. Y, Z, X, XY, ZY, Z2, ZX, X2 - Y2(output by shimmingtoolbox.coils.spherical_harmonics.spherical_harmonics), to
+        2. X, Y, Z, Z2, ZX, ZY, X2 - Y2, XY( in line with Siemens shims)
+        """
+
+        assert spher_harm.shape[3] == 8
+        reordered = np.zeros(spher_harm.shape)
+        reordered[:, :, :, 0] = spher_harm[:, :, :, 2]
+        reordered[:, :, :, 1] = spher_harm[:, :, :, 0]
+        reordered[:, :, :, 2] = spher_harm[:, :, :, 1]
+        reordered[:, :, :, 3] = spher_harm[:, :, :, 5]
+        reordered[:, :, :, 4] = spher_harm[:, :, :, 6]
+        reordered[:, :, :, 5] = spher_harm[:, :, :, 4]
+        reordered[:, :, :, 6] = spher_harm[:, :, :, 7]
+        reordered[:, :, :, 7] = spher_harm[:, :, :, 3]
+
+        return reordered
+
+    def get_scaling_factors():
+        """
+        Returns a vector of `scaling_factors` to apply to the(Siemens-reordered) 1st + 2nd order spherical harmonic fields
+        for rescaling field terms as "shim reference maps" in units of Hz/unit-shim:
+
+        Gx, Gy, and Gz should yield 1 micro-T of field shift per metre: equivalently, 0.042576 Hz/mm
+
+        2nd order terms should yield 1 micro-T of field shift per metre-squared: equivalently, 0.000042576 Hz/mm^2
+
+        Gist: given the stated nominal values, we can pick several arbitrary reference positions around the origin/isocenter
+        at which we know what the field *should* be, and use that to calculate the appropriate scaling factor.
+
+        NOTE: The method has been worked out empirically and has only been tested for 2 Siemens Prisma systems.
+        E.g.re: Y, Z, ZX, and XY terms, it was noted that their polarity had to be flipped relative to the form given by
```

Comment:
could take the active tense here to make it less wordy:
```suggestion
        E.g.re: Y, Z, ZX, and XY terms, their polarity had to be flipped relative to the form given by
```